gcc/rust/checks/errors/rust-unsafe-checker.cc

```diff
@@ -784,7 +784,25 @@ UnsafeChecker::visit (Trait &trait)
 void
 UnsafeChecker::visit (ImplBlock &impl)
 {
-  // FIXME: Handle unsafe impls
+  bool safe = !impl.is_unsafe ();
```

Comment:
I'm not sure if this is relevant, i might be talking nonsense, but i think there is something in the Unsafe class to detect unsafe context for you. You can use it like this to avoid populuting every new class that needs to be check for unsafe usage with a boolean unsafe

```
  if (unsafe_context.is_in_context ()) { /* code */ }

Reply:
These functions do exist, but I think we are in a different sort of "unsafe context" here that needs to be handled separately. an "unsafe impl" is used when implementing "unsafe traits", the functions inside the unsafe impl can be safe, and need "unsafe" contexts of their own to implement unsafe behavior.

From https://doc.rust-lang.org/reference/unsafe-keyword.html , "An unsafe trait is a trait that comes with extra safety conditions that must be upheld by implementations of the trait. The unsafe trait should come with documentation explaining what those extra safety conditions are. Such a trait must be prefixed with the keyword unsafe and can only be implemented by unsafe impl blocks."

Originally I tried doing what you suggested, however, the issue I ran into is that by entering the unsafe context, unsafe functions in unsafe impl's would no longer be required, which I think is out-of-spec.

As a final note, currently we don't enter an unsafe context when parsing unsafe impl's. I believe the "unsafe" token was simply ignored before I made changes to gcc/rust/hir/rust-ast-lower-item.cc.

Reply:
oh wow i didn't know there were multiple usage of unsafe, usually i thought there's only one way to use which is unsafe {}, thank you for your explanation

Reply:
@liamnaddell i'm curious, how did you find out if your `unsafe` boolean will be turned to true if there is an unsafe word before impl? Is a parse somewhere changing it ?

Reply:
I think the AST already had a get_unsafe (gcc/rust/hir/rust-ast-lower-item.cc:626). Then I checked using GDB and using the two tests I attached to the review